core/templates/dev/head/domain/topic_viewer/read-only-topic-object.factory.ts

```diff
@@ -0,0 +1,155 @@
+  'domain/topic/StoryReferenceObjectFactory';
+import { StorySummary } from
+  'domain/story/StorySummaryObjectFactory';
+import { Subtopic, SubtopicObjectFactory } from
+  'domain/topic/SubtopicObjectFactory';
+
+export interface ISkillDescriptions {
+  [skillId: string]: string | null;
+}
+
+export interface IDegreesOfMastery {
+  [skillId: string]: number | null;
+}
+
+export class ReadOnlyTopic {
+  _topicName: string;
+  _topicId: string;
+  _canonicalStories: Array<StorySummary>;
+  _additionalStories: Array<StorySummary>;
+  _uncategorizedSkills: Array<SkillSummary>;
+  _subtopics: Array<Subtopic>;
+  _degreesOfMastery: IDegreesOfMastery;
+  _skillDescriptions: ISkillDescriptions;
+
+  constructor(topicName: string, topicId: string,
+      canonicalStories: Array<StorySummary>,
+      additionalStories: Array<StorySummary>,
+      uncategorizedSkills: Array<SkillSummary>, subtopics: Array<Subtopic>,
+      degreesOfMastery: IDegreesOfMastery,
+      skillDescriptions: ISkillDescriptions) {
+    this._topicName = topicName;
+    this._topicId = topicId;
+    this._canonicalStories = canonicalStories;
+    this._additionalStories = additionalStories;
+    this._uncategorizedSkills = uncategorizedSkills;
+    this._subtopics = subtopics;
+    this._degreesOfMastery = degreesOfMastery;
+    this._skillDescriptions = skillDescriptions;
+  }
+
+  getTopicName(): string {
+    return this._topicName;
+  }
+
+  getTopicId(): string {
+    return this._topicId;
+  }
+
+  getCanonicalStories(): Array<StorySummary> {
+    return this._canonicalStories.slice();
+  }
+
+  getAdditionalStories(): Array<StorySummary> {
+    return this._additionalStories.slice();
+  }
+
+  getUncategorizedSkills(): Array<SkillSummary> {
+    return this._uncategorizedSkills.slice();
+  }
+
+  getSubtopics(): Array<Subtopic> {
+    return this._subtopics.slice();
+  }
+
+  getDegreesOfMastery(): IDegreesOfMastery {
+    return this._degreesOfMastery;
+  }
+
+  getSkillDescriptions(): ISkillDescriptions {
+    return this._skillDescriptions;
+  }
+}
+
+@Injectable({
+  providedIn: 'root'
+})
+
+export class ReadOnlyTopicObjectFactory {
+  _subtopicObjectFactory: SubtopicObjectFactory;
+  _skillSummaryObjectFactory: SkillSummaryObjectFactory;
+
+  constructor(
+    private subtopicObjectFactory: SubtopicObjectFactory,
+    private skillSummaryObjectFactory: SkillSummaryObjectFactory) {
+    this._subtopicObjectFactory = subtopicObjectFactory;
+    this._skillSummaryObjectFactory = skillSummaryObjectFactory;
+  }
+
+  getStorySummaryArray(storyDicts: any): Array<StorySummary> {
+    var storyReferenceObjectFactory = new StoryReferenceObjectFactory();
+    let storySummaryArray: Array<StorySummary> =
+    storyDicts.map((storyDict: any) => {
+      return new StorySummary( storyDict.id, storyDict.title,
+        storyDict.node_count, storyDict.description,
+        storyReferenceObjectFactory.createFromStoryId(storyDict.id).
```

Comment:
createFromStoryId() function always sets the published status to false, but here the published status is true, I think it should be fine to just pass 'true' here since in topic_viewer, we only return published stories from the backend.

Reply:
Where are we supposed to add `summary[published] = true` ?

Reply:
Somewhere after this [line ](https://github.com/oppia/oppia/blob/b36ede61274c4b06011925d31511b548f00193ef/core/controllers/topic_viewer.py#L80) in topic_viewer.py. Just iterate through all the stories and add the published = True, since earlier in the function, we only fetched the published stories, so it is fine to do this.
Then, you can use `storyDict.published` here directly.

Reply:
![image](https://user-images.githubusercontent.com/32016324/74462997-1fef9c80-4eb7-11ea-8bc7-daa832ebe7f5.png)
I think you were talking about something like this

Reply:
I did the changes, though `backend_tests` is not giving me 100% coverage. Says missing on 80-82 in `topic_viewer.py`. Should I remove `summary['published'] = True` for additional_story_dicts

Reply:
No, just modify topic_viewer_test, so you test those lines as well.

Reply:
Got delayed because of midsem. Actually so far wherever I've printed `additional_story_dicts`, its been an empty array. Also, `expected_dict` has it empty.